design/baremetal-operator/raid-api.md

```diff
@@ -0,0 +1,198 @@
+  one of `"1"`, `"0"`, or `"1+0"`.
+- The size for a volume is optional and should be expressed in
+  gibibytes.
+- The number of disks to use for the volume defaults to a minimum
+  value based on the RAID level.
+
+For software RAID, the inputs describe how to configure the operating
+system to see physical disks in a RAID configuration through software
+drivers, without using any special hardware.
+
+```yaml
+spec:
+  raid:
+    softwareVolumes:
+      - sizeGibibytes: 1000
+        level: 1
+        physicalDisks:
+          - deviceName: /dev/sda
+          - deviceName: /dev/sdb
+```
+
+- The RAID level for a volume is required and should be expressed as
+  one of `"1"`, `"0"`, or `"1+0"`.
+- The size for a volume is optional and should be expressed in
+  gibibytes.
+- The disks to use are specified using the same syntax as the
+  `rootDeviceHints`.
+
+Software RAID is constrained in other ways:
+
+- There must be 0, 1, or 2 software RAID devices.
+- The first software RAID volume must use RAID level 1 to ensure the
+  host boots in the event of a disk failure.
+- The second software RAID volume may use RAID level `0`, `1`, or
+  `1+0`.
+
+Regardless of the type of RAID used, the first volume defined is used
+as the root volume.
+
+The status fields of the `BareMetalHost` expand to include the RAID
+configuration used for the most recent provisioning operation.
+
+```yaml
+status:
+  provisioning:
+    raid:
+      hardwareVolumes: ...
+      softwareVolumes: ...
+```
+
+### Controller Changes
+
+Ironic applies RAID configuration during a cleaning phase, separate
+from the provisioning phase. In order for the metal3 controller to
+track this new phase, a new "preparing" state is introduced to the
+metal3 state machine, between "ready" and "provisioning". The
+controller and provisioner need to handle the new state.
+
+### Implementation Details/Notes/Constraints
+
+If the RAID instructions do not meet the validation criteria given
+above, the host is not provisioned and Ironic is not given any
+cleaning or provisioning instructions. The error is reported through
+the host status, and reconciliation stops until the settings are
+modified again.
+
+Valid RAID settings are stored in the status fields when provisioning
+starts, and only those values are used by the provisioner to ensure
+that if the spec fields are updated after provisioning has started
+Ironic is not given conflicting instructions. This is similar to how
+the root device hints are managed.
+
+### Risks and Mitigations
+
+Reconfiguring the RAID controller can lead to unexpected data loss, so
+during provisioning the existing configuration is only erased when
+explicit settings are given via the API. To manually configure a host,
+the user can pre-configure it and not provide RAID instructions to
+metal3.
+
+Leaving the storage of a host configured after it is deprovisioned can
+result in data exposure, so the host's RAID configuration is always
+erased during deprovisioning.
```

Comment:
This seems to conflict with the last sentence of the previous paragraph.

I'm also not sure this is true. Does the automated cleaning clean the physical disks, or just the logical ones?

Reply:
It's meant to be aspirational, so if the code does not do that now it should be changed to do that.

Should we say that the RAID settings are only erased if it looks like the BMO configured RAID on the host in the first place (i.e., there are RAID settings in the status fields)?

Reply:
The only time I am definitely sure that RAID settings are erased are next time you provision.

Reply:
> Should we say that the RAID settings are only erased if it looks like the BMO configured RAID on the host in the first place (i.e., there are RAID settings in the status fields)?

It would appear to be kind of artificial. The recommended approach would be to rely on Ironic for core functionality like this and only implement wrappers in BMO if it makes sense in a K8s environment. 

From what I know so far, Ironic can remove the ``logical_disks`` only if the ``delete_configuration`` clean/deploy step is used. In other words, it only removes ``logical_disks`` as of now. And I'm unsure if this would go so far as to zap the physical disks as well, I'm guessing it wouldn't, that looks like an added security measure that would take way too long. Imagine zapping 10TBs of spinners, and that's a highly optimistic number.

If we really desire this functionality of complete physical disk erase, Ironic would be a right place to put it, leaving BMO implementation unchanged, which would simply depend on Ironic to clean the physical disks in either the ``delete_configuration`` step or some other mechanism.

Reply:
A few comments, ...

1. Deletion of the existing RAID configuration can be requested during creation. That is expressed by `create_configuration`'s `delete_existing` parameter.
2. The `idrac` hardware type/driver does not zap the physical disks when it deletes the hardware RAID configuration. Other hardware types may behave the same.
3. Secure erase could be a means to quickly erase a large amount of storage.

Reply:
@dhellmann  If BMC supports RAID, `delete_configuration` step will done every time.

Reply:
> 1. Deletion of the existing RAID configuration can be requested during creation. That is expressed by `create_configuration`'s `delete_existing` parameter.

That would eliminate the need to explicitly invoke the ``delete_configuration`` step? am I assuming corrently?

> 2. The `idrac` hardware type/driver does not zap the physical disks when it deletes the hardware RAID configuration. Other hardware types may behave the same.

Just as I thought, makes sense, since it will take wayyyy too long and would not be desired in majority of use cases.

> 3. Secure erase could be a means to quickly erase a large amount of storage.


That could be looked into, as an alternative of disk zap, is we want to have this functionality, it's a good option.

Reply:
> > 1. Deletion of the existing RAID configuration can be requested during creation. That is expressed by `create_configuration`'s `delete_existing` parameter.
> 
> That would eliminate the need to explicitly invoke the `delete_configuration` step? am I assuming corrently?
> 
Digging deeper, the `delete_existing` parameter is an optional argument which was added relatively recently. While the `idrac` driver supports it, not all drivers do.

Reply:
> The only time I am definitely sure that RAID settings are erased are next time you provision.

The Ironic state machine can move from deleting to cleaning before going back to available or manageable. So, I think we can make Ironic wipe the RAID settings when it deprovisions the host.

Reply:
> A few comments, ...
> 
> 1. Deletion of the existing RAID configuration can be requested during creation. That is expressed by `create_configuration`'s `delete_existing` parameter.

Yes, the implementation will always start by deleting what is there. I think we should also delete those settings when deprovisioning a host.

> 2. The `idrac` hardware type/driver does not zap the physical disks when it deletes the hardware RAID configuration. Other hardware types may behave the same.
> 3. Secure erase could be a means to quickly erase a large amount of storage.

Sure, we can look at adding a secure wipe option as a separate enhancement.

Reply:
> > A few comments, ...
> > 
> > 1. Deletion of the existing RAID configuration can be requested during creation. That is expressed by `create_configuration`'s `delete_existing` parameter.
> 
> Yes, the implementation will always start by deleting what is there. I think we should also delete those settings when deprovisioning a host.

Do we still need to execute `delete_configuration` in `preparing`, if we execute `delete_configuration` in `deprovisioning`?

Reply:
I believe so. Since the node being provisioned for the first time will still need the ``delete_configuration`` in the ``preparing`` stage.

Reply:
Yes, that's right. We have to clean the host before we provision because we don't know what state it is in and we need to reset it. We have to clean the host when we deprovision it because we want to leave it in a blank state for the next use.

Reply:
AFAIK, the implementation always does a ``delete_configuration`` before ``apply_configuration``. And I'm sure it would also be doing a ``delete_configuration`` in deprovisioning. So we're covered here.

Reply:
> > The only time I am definitely sure that RAID settings are erased are next time you provision.
> 
> The Ironic state machine can move from deleting to cleaning before going back to available or manageable. So, I think we can make Ironic wipe the RAID settings when it deprovisions the host.

Yes, we could, by doing manual cleaning (although this would be a giant pain because we'd probably have to add a 'Cleaning' provisioning state after Deprovisioning). But currently we only do automatic cleaning, and its not clear to me that that will clean the physical disks (and thus wipe the RAID config off them) and not just the contents of the RAID volumes.

Reply:
The node will only be actually used in the "active" state, so we only need to execute `delete_configuration` clean step by `automated cleaning` in `deleting` state to avoid data exposure,  right?

Reply:
We also need to leave the node "cleaned" after deleting it. So we will need to remove all RAID volumes/disks (hardware/software) when node is deleted.

Reply:
And AFAIK, ``automated_cleaning`` doesn't do that for us.

Reply:
I'd be open to an ironic change to allow `delete_configuration` during automated cleaning.

Reply:
@dtantsur @zaneb Could we use manual cleaning during deprovisioning? Or should we just say that the RAID configuration is left intact after deprovisioning, for now?

Reply:
@dhellmann If we use `manual cleaning` during deprovisioning, we need to add  a new state. Will this be allowed?

Reply:
@Hellcatlk if we need a new state to keep track of what is happening, then yes, I would have no problem accepting it. We should look at how much work it would be to do what @dtantsur suggested, and allow `delete_configuration` during automated cleaning. That seems like it may be a better solution overall.

Reply:
I see, thank you.

Reply:
@dhellmann  I suggest, we can leave it intact for now and work on an incremental change later, in interest of time. Let's finish this and quickly merge the implementation PR as well, we have a couple more PRs coming up afterwards that will need review, and it is going to take a substantial amount of time. Really appreciate the expedition.

Reply:
@dtantsur PTAL

Reply:
I think https://review.opendev.org/744117 will actually be enough to be able to use raid.delete_configuration during cleaning.